cpp/src/arrow/csv/parser.cc

```diff
@@ -184,6 +184,31 @@ class SSE42Filter {
   const BulkFilterType filter_;
 };
 
+#elif defined ARROW_HAVE_NEON
+
+// NEON filter: 8 bytes at a time, comparing with all special chars
+
+class NeonFilter {
+ public:
+  using WordType = uint8x8_t;
+
+  explicit NeonFilter(const ParseOptions& options)
+      : delim_(vdup_n_u8(options.delimiter)),
+        quote_(vdup_n_u8(options.quoting ? options.quote_char : '\n')),
+        escape_(vdup_n_u8(options.escaping ? options.escape_char : '\n')) {}
+
+  bool Matches(WordType w) {
+    const uint8x8_t v = vceq_u8(w, vdup_n_u8('\r')) | vceq_u8(w, vdup_n_u8('\n')) |
+                        vceq_u8(w, delim_) | vceq_u8(w, quote_) | vceq_u8(w, escape_);
```

Comment:
Perhaps the filter classes should be templated with `SpecializedOptions`? This way, you would only check for quotes and escapes if needed.

Reply:
Good idea. Will try.

Reply:
See big improvement on Neoverse N1 after templating options.
No obvious difference for M1 (possibly due to good instruction parallelism).
No influence to sse4 and bloom filter.